File: frontend/src/types/booking.ts
import type { StepType } from "../lib/scheduleConfig";

export interface RejectedAttempt {
  slot_index: number;
  blocking_resource: string;
  offset: number;
}

export interface BookingSlot {
  slot_index: number;
  resource_type: StepType;
  resource_id?: string | null;
}

export interface BookingSearchResponse {
  pathway_id: string;
  date: string;
  earliest_start_slot: number | null;
  end_slot: number | null;
  feasible_starts: number[];
  rejected_attempts: RejectedAttempt[];
  slots: BookingSlot[];
  total_blocks: number;
}

export interface BookingSearchRequest {
  pathway_id: string;
  date: string;
}

export interface BookingCreateRequest {
  pathway_id: string;
  date: string;
  start_slot: number;
}

export interface Booking {
  id: string;
  patient_id: string;
  pathway_id: string;
  pathway_name?: string | null;
  date: string;
  start_slot: number;
  end_slot?: number | null;
  status: "confirmed" | "cancelled";
  created_at?: string | null;
  slots: BookingSlot[];
}

File: backend/app/schemas/booking.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.booking import BookingStatus
from app.models.pathway import StepResourceType


class BookingSearchRequest(BaseModel):
    pathway_id: UUID
    date: date


class RejectedAttemptOut(BaseModel):
    slot_index: int
    blocking_resource: str
    offset: int = 0


class BookingSlotOut(BaseModel):
    slot_index: int
    resource_type: StepResourceType
    resource_id: UUID | None = None


class BookingSearchResponse(BaseModel):
    pathway_id: UUID
    date: date
    earliest_start_slot: int | None
    end_slot: int | None
    feasible_starts: list[int] = []
    rejected_attempts: list[RejectedAttemptOut]
    slots: list[BookingSlotOut]
    total_blocks: int


class BookingCreateRequest(BaseModel):
    pathway_id: UUID
    date: date
    start_slot: int = Field(ge=0)


class BookingOut(BaseModel):
    id: UUID
    patient_id: UUID
    pathway_id: UUID
    pathway_name: str | None = None
    date: date
    start_slot: int
    end_slot: int | None = None
    status: BookingStatus
    created_at: datetime | None = None
    slots: list[BookingSlotOut] = []

    model_config = {"from_attributes": True}


class BookingConflictOut(BaseModel):
    detail: str
    suggestion: BookingSearchResponse | None = None
